lai_textclf/tensorboard.py

```diff
@@ -0,0 +1,132 @@
+from uuid import uuid4
+
+import lightning as L
+import concurrent.futures
+import os
+from pathlib import Path
+from time import time
+from typing import Optional
+
+
+from fsspec.implementations.local import LocalFileSystem
+from lightning.app.storage import Drive
+from lightning.app.storage.path import _filesystem
+from lightning.pytorch.utilities.rank_zero import rank_zero_only
+from lightning.app.utilities.cloud import is_running_in_cloud
+
+
+class DriveTensorBoardLogger(L.pytorch.loggers.TensorBoardLogger):
+    def __init__(self, *args, drive: Drive, refresh_time: int = 5, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.timestamp = None
+        self.drive = drive
+        self.refresh_time = refresh_time
+
+    @rank_zero_only
+    def log_metrics(self, metrics, step) -> None:
+        super().log_metrics(metrics, step)
+        if self.timestamp is None:
+            self._upload_to_storage()
+            self.timestamp = time()
+        elif (time() - self.timestamp) > self.refresh_time:
+            self._upload_to_storage()
+            self.timestamp = time()
+
+    def _upload_to_storage(self):
+        fs = _filesystem()
+        fs.invalidate_cache()
+
+        source_path = Path(self.log_dir).resolve()
+        destination_path = self.drive._to_shared_path(
+            self.log_dir, component_name=self.drive.component_name
+        )
+
+        def _copy(from_path: Path, to_path: Path) -> Optional[Exception]:
+
+            try:
+                # NOTE: S3 does not have a concept of directories, so we do not need to create one.
+                if isinstance(fs, LocalFileSystem):
+                    fs.makedirs(str(to_path.parent), exist_ok=True)
+
+                fs.put(str(from_path), str(to_path), recursive=False)
+
+                # Don't delete tensorboard logs.
+                if "events.out.tfevents" not in str(from_path):
+                    os.remove(str(from_path))
+
+            except Exception as e:
+                # Return the exception so that it can be handled in the main thread
+                return e
+
+        src = [file for file in source_path.rglob("*") if file.is_file()]
+        dst = [destination_path / file.relative_to(source_path) for file in src]
+
+        with concurrent.futures.ThreadPoolExecutor(4) as executor:
+            results = executor.map(_copy, src, dst)
+
+        # Raise the first exception found
+        exception = next((e for e in results if isinstance(e, Exception)), None)
+        if exception:
+            raise exception
+
+
+class TensorBoardWork(L.app.LightningWork):
+    def __init__(self, *args, drive: Drive, **kwargs):
+        super().__init__(
+            *args,
+            parallel=True,
+            cloud_build_config=L.BuildConfig(requirements=["tensorboard"]),
+            **kwargs,
+        )
+
+        self.drive = drive
+
+    def run(self):
+
+        use_localhost = not L.app.utilities.cloud.is_running_in_cloud()
```

Comment:
NIT: You are importing is_running_in_cloud two different ways here and below :)

Reply:
done, thanks :)